File: src/core/player.cpp
// SPDX-License-Identifier: GPL-3.0-or-later

#include "core/player.h"

Player::Player(QObject *parent)
    : QObject(parent), id(0), state(Player::Invalid), totalGameTime(0), ready(false),
    totalGames(0), winCount(0), runCount(0) {}

Player::~Player() {}

int Player::getId() const { return id; }

void Player::setId(int id) { this->id = id; }

QString Player::getScreenName() const { return screenName; }

void Player::setScreenName(const QString &name) {
  this->screenName = name;
  emit screenNameChanged();
}

QString Player::getAvatar() const { return avatar; }

void Player::setAvatar(const QString &avatar) {
  this->avatar = avatar;
  emit avatarChanged();
}

int Player::getTotalGameTime() const { return totalGameTime; }

void Player::addTotalGameTime(int toAdd) {
  totalGameTime += toAdd;
}

Player::State Player::getState() const { return state; }

QString Player::getStateString() const {
  switch (state) {
  case Online:
    return QStringLiteral("online");
  case Trust:
    return QStringLiteral("trust");
  case Run:
    return QStringLiteral("run");
  case Leave:
    return QStringLiteral("leave");
  case Robot:
    return QStringLiteral("robot");
  case Offline:
    return QStringLiteral("offline");
  default:
    return QStringLiteral("invalid");
  }
}

void Player::setState(Player::State state) {
  this->state = state;
  emit stateChanged();
}

void Player::setStateString(const QString &state) {
  if (state == QStringLiteral("online"))
    setState(Online);
  else if (state == QStringLiteral("trust"))
    setState(Trust);
  else if (state == QStringLiteral("run"))
    setState(Run);
  else if (state == QStringLiteral("robot"))
    setState(Robot);
  else if (state == QStringLiteral("offline"))
    setState(Offline);
  else
    setState(Invalid);
}

bool Player::isReady() const { return ready; }

void Player::setReady(bool ready) {
  this->ready = ready;
  emit readyChanged();
}

QList<int> Player::getGameData() {
  return QList<int>({ totalGames, winCount, runCount });
}

void Player::setGameData(int total, int win, int run) {
  totalGames = total;
  winCount = win;
  runCount = run;
  emit gameDataChanged();
}

QString Player::getLastGameMode() const {
  return lastGameMode;
}

void Player::setLastGameMode(const QString &mode) {
  lastGameMode = mode;
}

bool Player::isDied() const {
  return died;
}

void Player::setDied(bool died) {
  this->died = died;
}

File: lua/lsp/player.lua
-- SPDX-License-Identifier: GPL-3.0-or-later

---@meta

---@class fk.Player
---@field private id integer
---@field private screenName string
---@field private avatar string
---@field private state integer
---@field private died boolean
---@field public _fake_router fk.Client
local FPlayer = {}

---@return integer
function FPlayer:getId()
  return self.id
end

---@param id integer
function FPlayer:setId(id)
  self.id = id
end

---@return string
function FPlayer:getScreenName()
  return self.screenName
end

---@param name string
function FPlayer:setScreenName(name)
  self.screenName = name
end

---@return string
function FPlayer:getAvatar()
  return self.avatar
end

---@param avatar string
function FPlayer:setAvatar(avatar)
  self.avatar = avatar
end

---@return integer
function FPlayer:getTotalGameTime()
  return 0
end

---@param toAdd integer
function FPlayer:addTotalGameTime(toAdd) end

---@return integer
function FPlayer:getState()
  return self.state
end

---@param state integer
function FPlayer:setState(state)
  self.state = state
end

---@return integer[]
function FPlayer:getGameData()
end

---@param total integer
---@param win integer
---@param run integer
function FPlayer:setGameData(total, win, run) end

---@return boolean
function FPlayer:isDied()
  return self.died
end

---@param died boolean
function FPlayer:setDied(died)
  self.died = died
end

return FPlayer
